src/audio/pipewire.c

```diff
@@ -0,0 +1,318 @@
+    pw_thread_loop_lock(state->loop);
+    if (pw_stream_get_state(state->stream, NULL) == PW_STREAM_STATE_PAUSED)
+    {
+        pw_stream_set_active(state->stream, true);
+    }
+    pw_thread_loop_unlock(state->loop);
+    track_buffer_size = track.num_samples * state->stride; // we want bytes, not frames
+    // as long as we have samples in the buffer sent by speech dispatcher, we loop, blocking spd from calling us till the audio system played what we loaded in it
+    while (track_buffer_size > 0)
+    {
+        // drainage spinlock!
+        // we wait till fill_quantity is not empty after this thread can continue, aka after on_process updated that ringbuffer index with what it managed to play, so that we have some room in the buffer to fill from what's remaining of what was given to us by speech dispatcher
+        while (true)
+        {
+            fill_quantity = spa_ringbuffer_get_write_index(&state->rb, &write_index);
+            // see if there's some room in the buffer, if so, stop locking
+            room_left_in_buffer = SAMPLE_BUFFER_SIZE - fill_quantity;
+            if (room_left_in_buffer > 0)
+                // then we have room in the buffer, so stop locking our thread
+                break;
+            // otherwise, to not starve a cpu core of resources, we suspend our spinning until something, in this case on_process, wakes us up. In an embedded system without these primitives, spinning endlessly like that would be the only option. We could do the same here, but for efficiency reasons alone, we don't
+            pw_thread_loop_lock(state->loop);
+            struct pw_loop *program_loop = pw_thread_loop_get_loop(state->loop);
+            pw_thread_loop_unlock(state->loop);
+            spa_system_eventfd_read(program_loop->system, state->eventfd_number, &dummy);
+        }
+        // we write the amount of samples we can at this time without overflowing the buffer, to the memory area represented by the amount of room that's free after the last on_process call, to then be enqueued by pipewire
+        //  we assume speech dispatcher gives us the correct number of bytes for the format it chose, enough for this chunk. If that's not the case, there's not much we could do besides reading uninitialized memory or an incomplete chunk, unfortunately
+        // in case we produced more audio than the available room in the buffer that got freed by on_process, we only push what we can, otherwise we would read from uninitialised memory
+        room_left_in_buffer = SPA_MIN(track_buffer_size, room_left_in_buffer);
+        // everything has been accounted for, so write the first batch of data, keeping into account that we can't overflow the initially allocated block for the buffer, represented by SAMPLE_BUFFER_SIZE
+        spa_ringbuffer_write_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, write_index % SAMPLE_BUFFER_SIZE, track.samples, room_left_in_buffer);
```

Comment:
track.samples is incorrect here, it's always the same on each loop iteration. You need to use a variable that you advance